src/charm.py

```diff
@@ -379,25 +443,87 @@ def proceed_on_broken_event(self, event) -> bool:
+    def get_mongos_hosts_for_client(self) -> Set:
+        """Returns the hosts for mongos as a str.
+
+        The host for mongos can be either the K8s pod name or an IP address depending on how
+        the app has been configured.
+        """
+        if self.is_external_client:
+            return self.get_ext_mongos_hosts()
+
+        return self.get_k8s_mongos_hosts()
+
+    def get_k8s_mongos_hosts(self) -> Set:
+        """Returns the K8s hosts for mongos"""
+        hosts = set()
+        for unit in self.get_units():
+            hosts.add(self.get_k8s_mongos_host(unit))
+
+        return hosts
+
+    def get_ext_mongos_host(self, unit: Unit, incl_port=True) -> str | None:
+        """Returns the ext hosts for mongos on the provided unit."""
+        if not self.is_external_client:
+            return None
+
+        try:
+            unit_ip = self.node_port_manager.get_node_ip(unit.name)
+            if not incl_port:
+                return unit_ip
+
+            unit_port = self.node_port_manager.get_node_port(
+                port_to_match=Config.MONGOS_PORT, unit_name=unit.name
+            )
+            if unit_ip and unit_port:
+                return "{unit_ip}:{unit_port}"
+            else:
+                raise NoExternalHostError(f"No external host for unit {unit.name}")
+        except (
+            NoExternalHostError,
+            FailedToFindNodePortError,
+            FailedToFindServiceError,
+        ) as e:
+            raise FailedToGetHostsError(
+                "Failed to retrieve external hosts due to %s", e
+            )
+
+    def get_k8s_mongos_host(self, unit: Unit) -> str:
+        """Create a DNS name for a MongoDB unit.
+
+        Args:
+            unit_name: the juju unit name, e.g. "mongodb/1".
+
+        Returns:
+            A string representing the hostname of the MongoDB unit.
         """
-        return self.unit_host(self.unit)
+        unit_id = unit.name.split("/")[1]
+        return f"{self.app.name}-{unit_id}.{self.app.name}-endpoints"
+
+    def get_ext_mongos_hosts(self) -> Set:
+        """Returns the ext hosts for mongos.
+
+        Note: for external connections it is not enough to know the external ip, but also the
+        port that is associated with the client.
+        """
+        hosts = set()
+        for unit in self.get_units():
+            hosts.add(self.get_ext_mongos_host(unit))
 
-    def get_mongos_hosts(self) -> Set:
+        return hosts
+
+    def get_mongos_host(self) -> str:
         """Returns the host for mongos as a str.
 
         The host for mongos can be either the Unix Domain Socket or an IP address depending on how
         the client wishes to connect to mongos (inside Juju or outside).
         """
-        hosts = {self.unit_host(self.unit)}
-        for unit in self.peers_units:
-            hosts.add(self.unit_host(unit))
-
-        return hosts
+        return self.get_k8s_mongos_host(self.unit)
```

Comment:
**suggestion:** I'm not sure that all these functions are providing much value being broken up like this. It's also fairly confusing to read/search two very similar functions with only 1 char difference in the name.

From a quick search (I may have missed all of them fyi):
- You don't use `get_mongos_host()` and it just serves to call a different func, probably can be removed
- `get_k8s_mongos_host()` and `get_ext_mongos_host()` also both seem fairly redundant, as you only call it in the plural equivalent of the function. Might as well just put that logic there and remove the two functions

Reply:
You're not wrong this is super confusing. In part because of how some of the substrate agnostic libs need certain functions. I think if we need to change things we need up update the libs. If this is a strong suggestion let me know and I can update the lib code on the VM charm.

i.e. `get_mongos_host` is needed in `lib/charms/mongodb/v0/config_server_interface.py`

I can remove `get_k8s_mongos_host()` and `get_ext_mongos_host()` and consolidate those like you suggested though, that is an on point suggestion!

Reply:
It's a suggestion, so feel free to keep or discard, up to you 👍🏾